File: packages/dotnet/src/Webhooks/WebhookSignature.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Opensms.Internal;

namespace Opensms
{
    /// <summary>
    /// Verifies the <c>X-OpenSMS-Signature</c> header on webhook deliveries. Needs
    /// no API key, so a receiver can use it without constructing a client.
    /// </summary>
    /// <remarks>
    /// The header is <c>t=&lt;unix seconds&gt;,v1=&lt;hex&gt;</c> where <c>v1</c> is
    /// HMAC-SHA256 of <c>"&lt;t&gt;.&lt;raw body&gt;"</c> keyed with the full
    /// <c>whsec_...</c> secret as UTF-8 bytes. Always verify the exact raw bytes you
    /// received, before parsing JSON.
    /// </remarks>
    public static class WebhookSignature
    {
        /// <summary>The header carrying the signature.</summary>
        public const string HeaderName = "X-OpenSMS-Signature";

        /// <summary>Default clock-skew tolerance (300 seconds, inclusive).</summary>
        public static readonly TimeSpan DefaultTolerance = TimeSpan.FromSeconds(300);

        private enum Outcome { Valid, Invalid, Expired }

        /// <summary>True when <paramref name="header"/> is a valid, fresh signature of <paramref name="payload"/>.</summary>
        public static bool Verify(string payload, string? header, string? secret, TimeSpan? tolerance = null, DateTimeOffset? now = null)
            => Check(Encoding.UTF8.GetBytes(payload ?? ""), header, secret, tolerance, now) == Outcome.Valid;

        /// <summary>True when <paramref name="header"/> is a valid, fresh signature of the raw <paramref name="payload"/> bytes.</summary>
        public static bool Verify(byte[] payload, string? header, string? secret, TimeSpan? tolerance = null, DateTimeOffset? now = null)
            => Check(payload ?? Array.Empty<byte>(), header, secret, tolerance, now) == Outcome.Valid;

        /// <summary>
        /// Verify and parse a delivery. Throws <see cref="OpensmsException"/> with
        /// <c>Status == 0</c> and <c>Code</c> <c>invalid_signature</c> or
        /// <c>expired_signature</c> (or <c>invalid_payload</c> when the verified body is not an event).
        /// </summary>
        public static WebhookEvent ConstructEvent(string payload, string? header, string? secret, TimeSpan? tolerance = null, DateTimeOffset? now = null)
            => ConstructEvent(Encoding.UTF8.GetBytes(payload ?? ""), header, secret, tolerance, now);

        /// <inheritdoc cref="ConstructEvent(string, string?, string?, TimeSpan?, DateTimeOffset?)"/>
        public static WebhookEvent ConstructEvent(byte[] payload, string? header, string? secret, TimeSpan? tolerance = null, DateTimeOffset? now = null)
        {
            payload ??= Array.Empty<byte>();
            switch (Check(payload, header, secret, tolerance, now))
            {
                case Outcome.Expired:
                    throw new OpensmsException(0, "Webhook signature timestamp is outside the tolerance.", code: "expired_signature");
                case Outcome.Invalid:
                    throw new OpensmsException(0, "Webhook signature is invalid.", code: "invalid_signature");
            }
            try
            {
                return JsonSerializer.Deserialize<WebhookEvent>(payload, Json.Options)
                    ?? throw new JsonException("empty payload");
            }
            catch (JsonException ex)
            {
                throw new OpensmsException(0, "Webhook payload is not a valid event.", code: "invalid_payload", innerException: ex);
            }
        }

        /// <summary>Compute a header value the way the API does (useful for testing your receiver).</summary>
        public static string Sign(string secret, byte[] payload, DateTimeOffset at)
        {
            if (string.IsNullOrWhiteSpace(secret)) throw new ArgumentException("secret is required", nameof(secret));
            var t = at.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture);
            return "t=" + t + ",v1=" + Convert.ToHexString(Mac(secret, t, payload)).ToLowerInvariant();
        }

        private static Outcome Check(byte[] payload, string? header, string? secret, TimeSpan? tolerance, DateTimeOffset? now)
        {
            var tol = tolerance ?? DefaultTolerance;
            if (string.IsNullOrWhiteSpace(secret) || tol < TimeSpan.Zero || header == null) return Outcome.Invalid;

            var values = ParseHeader(header);
            if (values == null) return Outcome.Invalid;
            if (!long.TryParse(values["t"], NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture, out var ts))
                return Outcome.Invalid;

            var nowUnix = (now ?? DateTimeOffset.UtcNow).ToUnixTimeSeconds();
            var tolSeconds = (long)Math.Floor(tol.TotalSeconds);
            if (nowUnix - ts > tolSeconds || ts - nowUnix > tolSeconds) return Outcome.Expired;

            var v1 = values["v1"];
            if (v1.Length != 64) return Outcome.Invalid;
            byte[] want;
            try { want = Convert.FromHexString(v1); }
            catch (FormatException) { return Outcome.Invalid; }

            var got = Mac(secret!, values["t"], payload);
            return CryptographicOperations.FixedTimeEquals(got, want) ? Outcome.Valid : Outcome.Invalid;
        }

        /// <summary>Mirrors the server: comma separated, trimmed, split on the first '=', exactly the keys t and v1.</summary>
        private static Dictionary<string, string>? ParseHeader(string header)
        {
            var values = new Dictionary<string, string>(StringComparer.Ordinal);
            foreach (var part in header.Split(','))
            {
                var trimmed = part.Trim();
                var eq = trimmed.IndexOf('=');
                if (eq <= 0 || eq == trimmed.Length - 1) return null;
                var key = trimmed.Substring(0, eq);
                if (values.ContainsKey(key)) return null;
                values[key] = trimmed.Substring(eq + 1);
            }
            if (values.Count != 2 || !values.ContainsKey("t") || !values.ContainsKey("v1")) return null;
            return values;
        }

        private static byte[] Mac(string secret, string timestamp, byte[] body)
        {
            using var h = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
            var prefix = Encoding.UTF8.GetBytes(timestamp + ".");
            h.TransformBlock(prefix, 0, prefix.Length, null, 0);
            h.TransformFinalBlock(body, 0, body.Length);
            return h.Hash!;
        }
    }
}

File: packages/java/src/main/java/io/opensms/WebhookSignature.java
package io.opensms;

import io.opensms.models.WebhookEvent;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.HashMap;
import java.util.Map;

/**
 * Verifies {@code X-OpenSMS-Signature: t=<unix>,v1=<hex>} headers. {@code v1} is
 * {@code hex(HMAC-SHA256(secret, "<t>." + rawBody))}, keyed with the full
 * {@code whsec_...} secret as UTF-8 bytes. Mirrors the server's parser exactly.
 * Needs no API key, so a webhook receiver can use it on its own.
 *
 * <pre>{@code
 * WebhookEvent event = WebhookSignature.constructEvent(rawBody, request.getHeader("X-OpenSMS-Signature"), secret);
 * }</pre>
 */
public final class WebhookSignature {

    /** The header carrying the signature. */
    public static final String HEADER = "X-OpenSMS-Signature";
    /** Default clock tolerance in seconds. */
    public static final long DEFAULT_TOLERANCE_SECONDS = 300;

    private WebhookSignature() {
    }

    /** Outcome of a verification. */
    enum Result { VALID, INVALID, EXPIRED }

    /**
     * @param payload raw body.
     * @param header  header value.
     * @param secret  endpoint secret.
     * @return whether valid, with the default tolerance and the current time.
     */
    public static boolean verify(byte[] payload, String header, String secret) {
        return check(payload, header, secret, DEFAULT_TOLERANCE_SECONDS, now()) == Result.VALID;
    }

    /**
     * @param payload raw body (UTF-8).
     * @param header  header value.
     * @param secret  endpoint secret.
     * @return whether valid, with the default tolerance and the current time.
     */
    public static boolean verify(String payload, String header, String secret) {
        return verify(bytes(payload), header, secret);
    }

    /**
     * @param payload          raw body.
     * @param header           header value.
     * @param secret           endpoint secret.
     * @param toleranceSeconds allowed clock difference.
     * @param nowEpochSeconds  the current time in Unix seconds (injectable for tests).
     * @return whether valid.
     */
    public static boolean verify(byte[] payload, String header, String secret, long toleranceSeconds,
                                 long nowEpochSeconds) {
        return check(payload, header, secret, toleranceSeconds, nowEpochSeconds) == Result.VALID;
    }

    /**
     * @param payload          raw body (UTF-8).
     * @param header           header value.
     * @param secret           endpoint secret.
     * @param toleranceSeconds allowed clock difference.
     * @param nowEpochSeconds  the current time in Unix seconds.
     * @return whether valid.
     */
    public static boolean verify(String payload, String header, String secret, long toleranceSeconds,
                                 long nowEpochSeconds) {
        return verify(bytes(payload), header, secret, toleranceSeconds, nowEpochSeconds);
    }

    /**
     * Verify, then parse the event envelope.
     *
     * @param payload raw body.
     * @param header  header value.
     * @param secret  endpoint secret.
     * @return the event.
     * @throws OpensmsException status 0, code {@code invalid_signature} or {@code expired_signature}.
     */
    public static WebhookEvent constructEvent(byte[] payload, String header, String secret) {
        return constructEvent(payload, header, secret, DEFAULT_TOLERANCE_SECONDS, now());
    }

    /**
     * @param payload raw body (UTF-8).
     * @param header  header value.
     * @param secret  endpoint secret.
     * @return the event.
     */
    public static WebhookEvent constructEvent(String payload, String header, String secret) {
        return constructEvent(bytes(payload), header, secret);
    }

    /**
     * @param payload          raw body.
     * @param header           header value.
     * @param secret           endpoint secret.
     * @param toleranceSeconds allowed clock difference.
     * @param nowEpochSeconds  the current time in Unix seconds.
     * @return the event.
     */
    public static WebhookEvent constructEvent(byte[] payload, String header, String secret,
                                              long toleranceSeconds, long nowEpochSeconds) {
        Result r = check(payload, header, secret, toleranceSeconds, nowEpochSeconds);
        if (r == Result.EXPIRED) {
            throw OpensmsException.local("Webhook signature timestamp is outside the tolerance", "expired_signature", null);
        }
        if (r != Result.VALID) {
            throw OpensmsException.local("Webhook signature is invalid", "invalid_signature", null);
        }
        try {
            return Json.MAPPER.readValue(payload, WebhookEvent.class);
        } catch (Exception e) {
            throw OpensmsException.local("Webhook payload is not valid JSON: " + e.getMessage(), "invalid_payload", e);
        }
    }

    /**
     * @param payload          raw body (UTF-8).
     * @param header           header value.
     * @param secret           endpoint secret.
     * @param toleranceSeconds allowed clock difference.
     * @param nowEpochSeconds  the current time in Unix seconds.
     * @return the event.
     */
    public static WebhookEvent constructEvent(String payload, String header, String secret,
                                              long toleranceSeconds, long nowEpochSeconds) {
        return constructEvent(bytes(payload), header, secret, toleranceSeconds, nowEpochSeconds);
    }

    /**
     * Compute a header value, for tests of your own receiver.
     *
     * @param payload          raw body.
     * @param secret           endpoint secret.
     * @param timestampSeconds Unix seconds.
     * @return {@code t=<ts>,v1=<hex>}.
     */
    public static String sign(byte[] payload, String secret, long timestampSeconds) {
        String ts = Long.toString(timestampSeconds);
        return "t=" + ts + ",v1=" + hex(mac(secret, ts, payload));
    }

    static Result check(byte[] payload, String header, String secret, long tolerance, long now) {
        if (secret == null || secret.trim().isEmpty() || tolerance < 0 || header == null || payload == null) {
            return Result.INVALID;
        }
        Map<String, String> values = new HashMap<>();
        for (String part : header.split(",", -1)) {
            String p = part.trim();
            int eq = p.indexOf('=');
            if (eq <= 0 || eq == p.length() - 1) {
                return Result.INVALID;
            }
            String k = p.substring(0, eq);
            if (values.containsKey(k)) {
                return Result.INVALID;
            }
            values.put(k, p.substring(eq + 1));
        }
        if (values.size() != 2 || !values.containsKey("t") || !values.containsKey("v1")) {
            return Result.INVALID;
        }
        long ts;
        try {
            ts = Long.parseLong(values.get("t"));
        } catch (NumberFormatException e) {
            return Result.INVALID;
        }
        if (now - ts > tolerance || ts - now > tolerance) {
            return Result.EXPIRED;
        }
        byte[] want = unhex(values.get("v1"));
        if (want == null || want.length != 32) {
            return Result.INVALID;
        }
        byte[] got = mac(secret, values.get("t"), payload);
        return MessageDigest.isEqual(got, want) ? Result.VALID : Result.INVALID;
    }

    private static byte[] mac(String secret, String ts, byte[] body) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            mac.update(ts.getBytes(StandardCharsets.UTF_8));
            mac.update((byte) '.');
            mac.update(body);
            return mac.doFinal();
        } catch (Exception e) {
            throw new IllegalStateException("HmacSHA256 unavailable", e);
        }
    }

    private static String hex(byte[] b) {
        StringBuilder sb = new StringBuilder(b.length * 2);
        for (byte x : b) sb.append(String.format("%02x", x));
        return sb.toString();
    }

    private static byte[] unhex(String s) {
        if (s.length() % 2 != 0) return null;
        byte[] out = new byte[s.length() / 2];
        for (int i = 0; i < out.length; i++) {
            int hi = Character.digit(s.charAt(2 * i), 16);
            int lo = Character.digit(s.charAt(2 * i + 1), 16);
            if (hi < 0 || lo < 0) return null;
            out[i] = (byte) ((hi << 4) | lo);
        }
        return out;
    }

    private static byte[] bytes(String s) {
        return s == null ? null : s.getBytes(StandardCharsets.UTF_8);
    }

    private static long now() {
        return System.currentTimeMillis() / 1000L;
    }
}
